### PR TITLE
Add support for Azure

### LINKED ISSUE #17: Add code in subctl to use cloud-prepare API to deploy Azure

### PR DESCRIPTION
**Pull request by Jaanki**
Depends on https://github.com/submariner-io/cloud-prepare/pull/203

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

### CODE REVIEW

**Comment by Jaanki** [comment]
go.mod needs to be updated with the cloud-prepare commit id providing azure pkg

**Review by tpantelis — DISMISSED** [review]

File: pkg/cloud/azure/azure.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+	gwDeployer, err := azure.NewOcpGatewayDeployer(azureCloud, msDeployer, config.GWInstanceType)
+	if err != nil {
+		return status.Error(err, "Failed to initialize a GatewayDeployer config")
+	}
+
+	return function(azureCloud, gwDeployer, status)
+}
+
+func ReadFromFile(metadataFile string) (string, string, error) {
```
Comment (tpantelis) [inline pkg/cloud/azure/azure.go]: We should use the new [ReadMetadataFile function](https://github.com/submariner-io/subctl/pull/59) once it's merged.

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/cloud/azure/azure.go (hunk 1 above)
Comment (Jaanki) [reply pkg/cloud/azure/azure.go]: Done

File: pkg/cloud/azure/azure.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+func RunOn(restConfigProducer *restconfig.Producer, config *Config, status reporter.Interface,
+	function func(api.Cloud, api.GatewayDeployer, reporter.Interface) error,
+) error {
+	if config.OcpMetadataFile != "" {
+		var err error
+
+		config.InfraID, config.Region, err = readMetadataFile(config.OcpMetadataFile)
+		if err != nil {
+			return status.Error(err, "Failed to read AWS information from OCP metadata file %q", config.OcpMetadataFile)
```
Comment (tpantelis) [inline pkg/cloud/azure/azure.go]:
```suggestion
			return status.Error(err, "Failed to read Azure information from OCP metadata file %q", config.OcpMetadataFile)
```

File: pkg/cloud/azure/azure.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+		}
+	}
+
+	status.Start("Retrieving Azure credentials from your Azure authorization file")
+
+	err := os.Setenv("AZURE_AUTH_LOCATION", config.AuthFile)
+	if err != nil {
+		return status.Error(err, "Error locating authorization file")
```
Comment (tpantelis) [inline pkg/cloud/azure/azure.go]:
```suggestion
		return status.Error(err, "Unable to set AZURE_AUTH_LOCATION env variable")
```

File: pkg/cloud/azure/azure.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+func RunOn(restConfigProducer *restconfig.Producer, config *Config, status reporter.Interface,
+	function func(api.Cloud, api.GatewayDeployer, reporter.Interface) error,
+) error {
+	if config.OcpMetadataFile != "" {
+		var err error
+
+		config.InfraID, config.Region, err = readMetadataFile(config.OcpMetadataFile)
+		if err != nil {
+			return status.Error(err, "Failed to read AWS information from OCP metadata file %q", config.OcpMetadataFile)
+		}
```
Comment (tpantelis) [inline pkg/cloud/azure/azure.go:58]:
```suggestion
		}
		
		status.Success("Obtained infra ID %q and region %q from OCP metadata file %q", config.InfraID, config.Region, config.OcpMetadataFile)
```

File: pkg/cloud/azure/azure.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	status.Start("Retrieving Azure credentials from your Azure authorization file")
```
Comment (tpantelis) [inline pkg/cloud/azure/azure.go]:
```suggestion
	status.Start("Retrieving Azure credentials from your Azure authorization file %q", config.AuthFile)
```

File: pkg/cloud/azure/azure.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	subscriptionID, err := initializeFromAuthFile(config.AuthFile)
+	if err != nil {
+		return status.Error(err, "Failed to read authorization information from Azure authorization file")
+	}
+
+	status.End()
+
+	status.Start("Initializing AWS connectivity")
```
Comment (tpantelis) [inline pkg/cloud/azure/azure.go]:
```suggestion
	status.Start("Initializing Azure connectivity")
```

File: pkg/cloud/azure/azure.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+		TenantID       string
+	}
+
+	err = json.Unmarshal(data, &authInfo)
+	if err != nil {
+		return "", errors.Wrap(err, "error unmarshalling data")
+	}
+
+	if err := os.Setenv("AZURE_CLIENT_ID", authInfo.ClientID); err != nil {
+		return "", errors.Wrapf(err, "Unable to set Azure Client ID env variable")
```
Comment (tpantelis) [inline pkg/cloud/azure/azure.go]:
```suggestion
		return "", errors.Wrapf(err, "unable to set AZURE_CLIENT_ID env variable")
```

File: pkg/cloud/azure/azure.go [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	if err := os.Setenv("AZURE_CLIENT_SECRET", authInfo.ClientSecret); err != nil {
+		return "", errors.Wrapf(err, "Unable to set Azure Client Secret env variable")
```
Comment (tpantelis) [inline pkg/cloud/azure/azure.go]:
```suggestion
		return "", errors.Wrapf(err, "Unable to set AZURE_CLIENT_SECRET env variable")
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/cloud/prepare/azure.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/reporter"
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+	"github.com/submariner-io/subctl/internal/restconfig"
+	"github.com/submariner-io/subctl/pkg/cloud"
+	"github.com/submariner-io/subctl/pkg/cloud/azure"
+)
+
+func Azure(restConfigProducer *restconfig.Producer, ports *cloud.Ports, config *azure.Config, status reporter.Interface) error {
+	status.Start("Preparing AWS cloud for Submariner deployment")
```
Comment (aswinsuryan) [inline pkg/cloud/prepare/azure.go]:
```suggestion
	status.Start("Preparing Azure cloud for Submariner deployment")
```

File: cmd/subctl/azure.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+		command.Flags().StringVar(&azureConfig.InfraID, infraIDFlag, "", "Azure infra ID")
+		command.Flags().StringVar(&azureConfig.Region, regionFlag, "", "Azure region")
+		command.Flags().StringVar(&azureConfig.OcpMetadataFile, "ocp-metadata", "",
+			"OCP metadata.json file (or directory containing it) to read Azure infra ID and region from (Takes precedence over the flags)")
+		command.Flags().StringVar(&azureConfig.AuthFile, "auth-file", "", "Azure authorization file to be used")
+	}
+
+	addGeneralAzureFlags(azurePrepareCmd)
+	azurePrepareCmd.Flags().IntVar(&azureConfig.Gateways, "gateways", defaultNumGateways, "Number of gateways to deploy")
+	// `Standard_F4s_v2` matches the most to `cd5.large` of AWS.
```
Comment (aswinsuryan) [inline cmd/subctl/azure.go:68]:
```suggestion
	// `Standard_F4s_v2` matches the most to `cd5.large` of Azure.
```

Comment (Jaanki) [reply cmd/subctl/azure.go:68]: It states that `cd5.large` type of AWS. So this change is not needed.

**Review by tpantelis — APPROVED** [review]

**Review by aswinsuryan — APPROVED** [review]

**Comment by nyechiel** [comment]
@Jaanki @aswinsuryan not sure if it's tracked separately, but we need to update the docs as well.

**Review by dfarrell07 — APPROVED** [review]

**Comment by aswinsuryan** [comment]
> @Jaanki @aswinsuryan not sure if it's tracked separately, but we need to update the docs as well.

@nyechiel  Created an issue https://github.com/submariner-io/submariner-website/issues/766

**Pull request merged**